### PR TITLE
Fix broken index migration

### PR DESCRIPTION
**Pull request by umanwizard**
Migrating indexes involves loading their stored SQL and re-parsing it. However, `is_retained_metrics_relation` and `custom_logical_compaction_window` are not expressed in SQL, and were being filled in as `false`/`None` by default. This was causing the retained metrics indexes to be created with 1s retention time, thus losing us the ability to show historical data.

I'm not sure if there's a good way to test this -- do we have a framework that will let us spin up an environment on v0.48.5, upgrade it to v0.49.2 (to observe the bad behavior), then upgrade it to this PR (to observe that the indexes are now retained) ?

If not, we will have to rely on the "I tried it and it works" method of testing.

### Motivation
  * This PR fixes a recognized bug.
https://github.com/MaterializeInc/materialize/issues/18699

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Fix lack of historical data in cluster replica metrics

### CODE REVIEW

File: src/adapter/src/catalog.rs [hunk 1]
```diff
@@ -2133,12 +2133,28 @@ impl CatalogItem {
         }
     }
 
-    pub fn initial_logical_compaction_window(&self) -> Option<Duration> {
-        let custom_logical_compaction_window = match self {
+    pub fn custom_logical_compaction_window(&self) -> Option<Duration> {
```
Comment (jkosh44) [inline src/adapter/src/catalog.rs:2136]: Can you add doc-comments to these two functions. It's not immediately obvious what the difference is.

File: src/adapter/src/catalog.rs [hunk 2]
```diff
@@ -2853,14 +2892,9 @@ impl Catalog {
...
                         });
-                    let CatalogItem::Index(catalog_index) = &mut item else {
+                    let CatalogItem::Index(_) = &mut item else {
                         panic!("internal error: builtin index {}'s SQL does not begin with \"CREATE INDEX\".", index.name);
                     };
-                    if index.is_retained_metrics_object {
-                        catalog_index.is_retained_metrics_object = true;
-                        catalog_index.custom_logical_compaction_window =
-                            Some(catalog.state.system_config().metrics_retention());
-                    }
```
Comment (jkosh44) [inline src/adapter/src/catalog.rs:2863]: Why does this get removed?

**Review by jkosh44 — APPROVED** [review]
LGTM. Can we add a test to the upgrade framework to make sure that this is fixed?

**Comment by jkosh44** [comment]
FYI @parker-timmerman, tagging you since you're working on migrations.

File: src/adapter/src/catalog.rs (hunk 2 above)
Comment (umanwizard) [reply src/adapter/src/catalog.rs:2863]: Because these are now set by the `parse_item` function, which is called above.

**Comment by umanwizard** [comment]
@jkosh44 can you give me a pointer to where said upgrade framework tests live?

**Comment by jkosh44** [comment]
It lives here: https://github.com/MaterializeInc/materialize/tree/main/test/upgrade

Phillip might be the right person to ask if you have any questions about it (I didn't explicitly tag him to avoid bothering him in case you don't have any questions).

**Comment by umanwizard** [comment]
I'm going to optimistically land this now; if Philip tells me it's indeed possible to add a test then I'll do so tomorrow.

**Pull request merged**